### PR TITLE
Overlay clipping

### PR DESCRIPTION
**Pull request by davidpiuva**
Culling away pixels that are outside of the target image when blitting images.

### CODE REVIEW

File: test/draw/gpu/BlitTest.ooc [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+BlitTest: class extends Fixture {
+	sourceImage := RasterMonochrome fromAscii(
+		"< *X>
+		<XXXXXX  >
+		<X*  *X  >
+		<X*  *X  >
+		<X* XXX  >
+		<X**XXX  >
+		<XXXXXX  >"
```
Comment (davidpiuva) [inline test/draw/gpu/BlitTest.ooc]: Both input image and output image has to be aligned for the GPU rendering to work. There might be an incorrect stride assumption in the GPU image constructors.

File: source/draw/RasterMonochrome.ooc [hunk 2]
```diff
@@ -64,17 +64,22 @@ RasterMonochrome: class extends RasterPacked {
...
+		startDst := offset toIntPoint2D()
+		endDst := startDst + sourceBox size
+		if (startDst x < 0) {
+			startSrc x -= startDst x
+			startDst x = 0
+		}
+		if (startDst y < 0) {
+			startSrc y -= startDst y
+			startDst y = 0
+		}
```
Comment (davidpiuva) [inline source/draw/RasterMonochrome.ooc]: Clip against upper left.

File: source/draw/RasterMonochrome.ooc [hunk 3]
```diff
@@ -64,17 +64,22 @@ RasterMonochrome: class extends RasterPacked {
...
+		endDst = endDst minimum(this size toIntPoint2D())
```
Comment (davidpiuva) [inline source/draw/RasterMonochrome.ooc]: Clip against bottom right.

File: source/draw/RasterMonochrome.ooc [hunk 4]
```diff
@@ -64,17 +64,22 @@ RasterMonochrome: class extends RasterPacked {
...
+		dimensions := endDst - startDst
+		endSrc := startSrc + dimensions
+		if (startSrc x >= 0 && startSrc y >= 0 && endSrc x <= source size x && endSrc y <= source size y && startDst x >= 0 && startDst y >= 0 && endDst x <= this size x && endDst y <= this size y) {
```
Comment (davidpiuva) [inline source/draw/RasterMonochrome.ooc]: Assert memory bounds.

**Comment by davidpiuva** [comment]
@christopherlagerhult Peer review?

File: test/draw/gpu/BlitTest.ooc [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+		(resultCpu, resultGpu, resultCpuFromGpu) free()
+	}
+	free: override func {
+		this sourceImage free()
+		super()
+	}
+}
+gpuContext := OpenGLContext new()
+BlitTest new() run() . free()
+gpuContext free()
```
Comment (christopherlagerhult) [inline test/draw/gpu/BlitTest.ooc]: Isn't this possible to place inside the class?

**Comment by christopherlagerhult** [comment]
Labeling this as final, but I would still like an answer to my question above :)

File: test/draw/gpu/BlitTest.ooc (hunk 5 above)
Comment (davidpiuva) [inline test/draw/gpu/BlitTest.ooc]: It is probably possible too.

**Pull request merged**